src/libutil/logging.cc

```diff
@@ -273,11 +303,57 @@ struct JSONLogger : Logger {
         addFields(json, fields);
         write(json);
     }
+
+    void result(ActivityId act, ResultType type, const nlohmann::json & j) override
+    {
+        nlohmann::json json;
+        json["action"] = "result";
+        json["id"] = act;
+        json["type"] = type;
+        json["payload"] = j;
+        write(json);
+    }
 };
 
-std::unique_ptr<Logger> makeJSONLogger(Descriptor fd)
+std::unique_ptr<Logger> makeJSONLogger(Descriptor fd, bool includeNixPrefix)
+{
+    return std::make_unique<JSONLogger>(fd, includeNixPrefix);
+}
+
+std::unique_ptr<Logger> makeJSONLogger(const std::filesystem::path & path, bool includeNixPrefix)
 {
-    return std::make_unique<JSONLogger>(fd);
+    struct JSONFileLogger : JSONLogger {
+        AutoCloseFD fd;
+
+        JSONFileLogger(AutoCloseFD && fd, bool includeNixPrefix)
+            : JSONLogger(fd.get(), includeNixPrefix)
+            , fd(std::move(fd))
+        { }
+    };
+
+    AutoCloseFD fd =
+        std::filesystem::is_socket(path)
+        ? connect(path)
+        : toDescriptor(open(path.c_str(), O_CREAT | O_APPEND | O_WRONLY, 0644));
+    if (!fd)
+        throw SysError("opening log file %1%", path);
+
+    return std::make_unique<JSONFileLogger>(std::move(fd), includeNixPrefix);
+}
+
+void applyJSONLogger()
+{
+    if (!loggerSettings.jsonLogPath.get().empty()) {
+        std::vector<std::unique_ptr<Logger>> loggers;
+        try {
+            loggers.push_back(std::move(logger));
+            loggers.push_back(makeJSONLogger(std::filesystem::path(loggerSettings.jsonLogPath.get()), false));
+            logger = makeTeeLogger(std::move(loggers));
+        } catch (...) {
+            logger = std::move(loggers[0]);
+            ignoreExceptionExceptInterrupt();
+        }
```

Comment:
(Standard caveat applies)

The catch branch unconditionally accessing `loggers[0]`, where `loggers` is only populated inside the `try` branch kinda scares me... Even though I'm mostly certain that `loggers.push_back(std::move(logger))` will never actually fail to insert into the vector (so it _should_ always be there)...

I'm imagining that the thing we really care about here is catching a potential `makeJSONLogger` or `makeTeeLogger` failure? Would it be worth it to initialize the `std::vector` with the `std::move(logger)`  immediately, instead of `push_back`ing it? Or even `push_back`ing it outside of the try-catch?

Reply:
> Would it be worth it to initialize the std::vector with the std::move(logger) immediately,

I tried that initially but it gave some rather confusing error messages wrt move semantics that I couldn't solve. Hence the `push_back()`. But the initializer list construction has the same problem with not being atomic.